### PR TITLE
Defining the Master Key Provider Keyring

### PR DESCRIPTION
**Pull request by WesleyRosenblum**
*Issue #, if available:*

*Description of changes:*

Initial definition of the Master Key Provider Keyring.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: framework/master-key-provider-keyring.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+0.1.0-preview
+
+## Implementations
+
+- [Java](https://github.com/aws/aws-encryption-sdk-java/blob/master/src/main/java/com/amazonaws/encryptionsdk/keyrings/MasterKeyProviderKeyring.java)
+- [Python](https://github.com/aws/aws-encryption-sdk-python/src/aws_encryption_sdk/keyrings/master_key.py)
+
+## Overview
+
+A keyring which wraps a legacy [Master Key Provider](master-key-provider-interface.md) to facilitate transitioning to [Keyrings](./keyring-interface.md).
```
Comment (mattsb42-aws) [inline framework/master-key-provider-keyring.md]:
```suggestion
A keyring that wraps a legacy [Master Key Provider](master-key-provider-interface.md) to facilitate transitioning to [Keyrings](./keyring-interface.md).
```

File: framework/master-key-provider-keyring.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+## Inputs
+
+On keyring initialization, a keyring MUST define the following:
+
+- [Master Key Provider](#master-key-provider)
+
+### Master Key Provider
+
+A [master key provider](master-key-provider-interface.md) that can decrypt data keys using available [master keys](master-key-interface.md)
+ as well as provide [master keys](master-key-interface.md) for use in encryption.
```
Comment (mattsb42-aws) [inline framework/master-key-provider-keyring.md]: I would leave the definition of a MKP to the MKP spec. Notably, this definition implies that the MKP provided MUST work on both encrypt and decrypt paths, which might not always be correct.

File: framework/master-key-provider-keyring.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+## Operation
+
+### OnEncrypt
+
+This keyring MUST acquire a list of [master keys](master-key-interface.md) by calling the master key provider's [Get Master Keys for Encryption](master-key-provider-interface.md#get-master-keys-for-encryption)
```
Comment (mattsb42-aws) [inline framework/master-key-provider-keyring.md]: [NP/Formatting] This is a pretty long line. I don't think we have a max-length rule for this repo yet but my general rule of thumb is semantic newlines and stay below ~90 characters. (/me makes a note to go looking for markdown linters again).

File: framework/master-key-provider-keyring.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+using the [encryption context](structures.md#encryption-context) from the [encryption materials](structures.md#encryption-materials) as input. If the 
+master key provider does not provide any master keys, OnEncrypt MUST fail.
+
+If the input [encryption materials](structures.md#encryption-materials) do not contain a [plaintext data key](structures.md#plaintext-data-key), OnEncrypt MUST
+attempt to generate a new plaintext data key and encrypt that data key by calling [Generate Data Key](master-key-interface.md#generate-data-key)
+on the first (primary) [master key](master-key-interface.md) provided by the master key provider.
```
Comment (mattsb42-aws) [inline framework/master-key-provider-keyring.md]: Again, leave this to the MKP spec.

From MKP#Get_Master_Keys_For_Encryption:
> The output of this interface MUST indicate which one of those master keys MUST be used to generate the data key.

File: framework/master-key-provider-keyring.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+- set the plaintext data key on the [encryption materials](structures.md#encryption-materials)
+- append a new [encrypted data key](structures.md#encrypted-data-key) to the encrypted data key list
+  in the [encryption materials](structures.md#encryption-materials), constructed as follows:
+  - the [ciphertext](structures.md#ciphertext) is the response encrypted data key.
+  - the [key provider id](structures.md#key-provider-id) is the provider id of the primary master key.
+  - the [key provider information](structures.md#key-provider-information) is the key id of the primary master key.
+- append a new record to the [keyring trace](structures.md#keyring-trace) 
+  in the input [encryption materials](structures.md#encryption-materials), constructed as follows:
+  - The string field KeyNamespace contains the provider id of the primary master key.
+  - The string field KeyName contains the key id of the primary master key.
```
Comment (mattsb42-aws) [inline framework/master-key-provider-keyring.md]: Types and specific names are Java-specific. Be generic.

File: framework/master-key-provider-keyring.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+If the call to Generate Data Key does not succeed, OnEncrypt MUST NOT modify the encryption materials and MUST fail.
+
+If the call succeeds, OnEncrypt MUST do the following with the response from [Generate Data Key](master-key-interface.md#generate-data-key)
+- set the plaintext data key on the [encryption materials](structures.md#encryption-materials)
+- append a new [encrypted data key](structures.md#encrypted-data-key) to the encrypted data key list
+  in the [encryption materials](structures.md#encryption-materials), constructed as follows:
+  - the [ciphertext](structures.md#ciphertext) is the response encrypted data key.
+  - the [key provider id](structures.md#key-provider-id) is the provider id of the primary master key.
+  - the [key provider information](structures.md#key-provider-information) is the key id of the primary master key.
```
Comment (mattsb42-aws) [inline framework/master-key-provider-keyring.md]: We should add an exception pointer here that refers to the Raw AES Keyring for how `RawMasterKey` and `JCEMasterKey` differ from this.

File: framework/master-key-provider-keyring.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+- append a new record to the [keyring trace](structures.md#keyring-trace) 
+  in the input [encryption materials](structures.md#encryption-materials), constructed as follows:
+  - The string field KeyNamespace contains the provider id of the primary master key.
+  - The string field KeyName contains the key id of the primary master key.
+  - The [flags](structures.md#flags) field of this record includes the following flags:
+    - [GENERATED DATA KEY](structures.md#flags)
+    - [ENCRYPTED DATA KEY](structures.md#flags)
+    - [SIGNED ENCRYPTION CONTEXT](structures.md#flags) (if the master key supports encryption context signing)
```
Comment (mattsb42-aws) [inline framework/master-key-provider-keyring.md]: How do I know if it does or does not?

File: framework/master-key-provider-keyring.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+Next, for each [master key](master-key-interface.md) provided by the master key provider's
```
Comment (mattsb42-aws) [inline framework/master-key-provider-keyring.md]:
```suggestion
For each remaining [master key](master-key-interface.md) provided by
```

File: framework/master-key-provider-keyring.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+  - The string field KeyName contains the key id of the primary master key.
+  - The [flags](structures.md#flags) field of this record includes the following flags:
+    - [GENERATED DATA KEY](structures.md#flags)
+    - [ENCRYPTED DATA KEY](structures.md#flags)
+    - [SIGNED ENCRYPTION CONTEXT](structures.md#flags) (if the master key supports encryption context signing)
+
+Next, for each [master key](master-key-interface.md) provided by the master key provider's 
+[Get Master Keys for Encryption](master-key-provider-interface.md#get-master-keys-for-encryption), the keyring MUST call
+ [EncryptDataKey](master-key-interface.md#encrypt-data-key). If the primary master key was used to generate a data key,
+ that master key MUST NOT be used again for encryption.
```
Comment (mattsb42-aws) [inline framework/master-key-provider-keyring.md]: I *think* this should be a SHOULD not a MUST. I'm open to being wrong though.

```suggestion
 that master key SHOULD NOT be used again for encryption.
```

File: framework/master-key-provider-keyring.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    - [ENCRYPTED DATA KEY](structures.md#flags)
+    - [SIGNED ENCRYPTION CONTEXT](structures.md#flags) (if the master key supports encryption context signing)
+
+### OnDecrypt
+
+If the input [decryption materials](structures.md#decryption-materials) contain a plaintext data key,
+OnDecrypt MUST immediately return the unmodified decryption materials.
+
+Otherwise, OnDecrypt MUST attempt to decrypt the [encrypted data keys](structures.md#encrypted-data-key)
+in the input using [Decrypt Data Key](master-key-provider-interface.md#decrypt-data-key)
```
Comment (mattsb42-aws) [inline framework/master-key-provider-keyring.md]: They also MUST be attempted in order.

File: framework/master-key-provider-keyring.md [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+If OnDecrypt fails to successfully decrypt the [encrypted data keys](structures.md#encrypted-data-key),
+then OnDecrypt MUST output the unmodified input [decryption materials](structures.md#decryption-materials).
+
+Otherwise, OnDecrypt MUST do the following with the response from [Decrypt Data Key](master-key-provider-interface.md#decrypt-data-key)
+- set the plaintext data key on the [decryption materials](structures.md#decryption-materials)
+- append a new record to the [keyring trace](structures.md#keyring-trace)
+  in the input [encryption materials](structures.md#encryption-materials), constructed as follows:
+  - The string field KeyNamespace contains the provider id of the master key that was used to decrypt the data key.
+  - The string field KeyName contains the key id of the master key that was used to decrypt the data key.
```
Comment (mattsb42-aws) [inline framework/master-key-provider-keyring.md]: Ditto type and names being Java-specific.

File: framework/master-key-provider-keyring.md (hunk 3 above)
Comment (WesleyRosenblum) [reply framework/master-key-provider-keyring.md]: Looking at the other documents in the spec, it looks like ~120 characters was generally the limit being followed, which I think is still reasonable. This line is over that even, I'll break it up

File: framework/master-key-provider-keyring.md (hunk 4 above)
Comment (WesleyRosenblum) [reply framework/master-key-provider-keyring.md]: Ah ok, the Java ESDK doesn't do that

File: framework/master-key-provider-keyring.md (hunk 9 above)
Comment (WesleyRosenblum) [reply framework/master-key-provider-keyring.md]: I would prefer to default to the more strict MUST NOT, I can't think of why we would want the master key to be used twice and SHOULD NOT leaves the door open for that.

File: framework/master-key-provider-keyring.md (hunk 7 above)
Comment (WesleyRosenblum) [reply framework/master-key-provider-keyring.md]: I don't think there is any specified way to determine that, other than knowing that KMS and Raw AES support it. I could change it to say (if the master key is known to support encryption context signing, such as the KMS Master Key and Raw AES Master Key)

File: framework/master-key-provider-keyring.md (hunk 2 above)
Comment (WesleyRosenblum) [reply framework/master-key-provider-keyring.md]: I've removed the definition

File: framework/master-key-provider-keyring.md (hunk 10 above)
Comment (WesleyRosenblum) [reply framework/master-key-provider-keyring.md]: I've specified now that the input to DecryptDataKey must contain the Encrypted Data Keys in their original order

File: framework/master-key-provider-keyring.md (hunk 5 above)
Comment (WesleyRosenblum) [reply framework/master-key-provider-keyring.md]: Updated

File: framework/master-key-provider-keyring.md (hunk 6 above)
Comment (WesleyRosenblum) [reply framework/master-key-provider-keyring.md]: Now it says: the key provider information is the key id of the master key that generated the key, with the exception of the raw AES master key. For a detailed description of this exception, see the [Raw AES Keyring specification]

File: framework/master-key-provider-keyring.md (hunk 11 above)
Comment (WesleyRosenblum) [reply framework/master-key-provider-keyring.md]: Updated

**Comment by WesleyRosenblum** [comment]
The Java ESDK cannot use the MasterKeyProviderKeyring internally due to backward compatibility issues, leaving only the Python ESDK as the potential implementation of this spec. Since there is only one language that would implement this, it doesn't need to be in the spec anymore so I'm closing this.

**Pull request closed without merging**